R/string_boundary_linter.R

```diff
@@ -0,0 +1,86 @@
+#' Require usage of startsWith() and endsWith() over grepl()/substr() versions
+#'
+#' [startsWith()] is used to detect fixed initial substrings; it is more
+#'   readable and more efficient than equivalents using [grepl()] or [substr()].
+#'   c.f. `startsWith(x, "abc")`, `grepl("^abc", x)`,
+#'   `substr(x, 1L, 3L) == "abc"`.
+#'
+#' Ditto for using [endsWith()] to detect fixed terminal substrings.
+#'
+#' @evalRd rd_tags("string_boundary_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+string_boundary_linter <- function() {
+  str_cond <- xp_and(
+    "string-length(text()) > 3",
+    "contains(text(), '^') or contains(text(), '$')"
+  )
+  grepl_xpath <- glue::glue("//expr[
+    expr[SYMBOL_FUNCTION_CALL[text() = 'grepl']]
```

Comment:
should be `expr[1]`

Reply:
which one, `expr[1]`? any reason? we don't do that anywhere else AFAIK -- `SYMBOL_FUNCTION_CALL` can only be in `position()=1`

Reply:
Exactly, `expr[1]/SYMBOL_FUNCTION_CALL`, for consistency, and maybe marginally improved performance.

Reply:
I think consistency here means not doing that -- we don't do it on any other linters AFAIK. Not strictly opposed to doing it universally but I suspect the improvement to be minimal. Happy to review a clean-up PR lintr-wide if you'd like.

Reply:
Okay, fine.
Might save a few ms especially on calls with many arguments, e.g. `list(...)`. But let's revisit this across lintr.

IMO we should specify the index by default unless we know the match might be in different locations.

Reply:
Agreed, just don't have the willpower to do another 70-linter cleanup at the moment :)